EKF/vel_pos_fusion.cpp

```diff
@@ -45,276 +45,175 @@
-	}
+	// vertical position
+	if(fuse_mask[VPOS]){
+		innov_var[5] = P[9][9] + obs_var[5];
+		test_ratio[VPOS] = sq(innov[5]) / (sq(innov_gate[VPOS]) * innov_var[5]);
 
-	_fuse_hor_vel = _fuse_hor_vel_aux = _fuse_vert_vel = false;
+		bool innov_check_pass = (test_ratio[VPOS] <= 1.0f) || !_control_status.flags.tilt_align;
+		if (innov_check_pass) {
+			_time_last_hgt_fuse = _time_last_imu;
+			_innov_check_fail_status.flags.reject_pos_D = false;
 
-	// record the successful position fusion event
-	if (pos_check_pass && _fuse_pos) {
-		if (!_fuse_hpos_as_odom) {
-			_time_last_pos_fuse = _time_last_imu;
+			// fuse the horizontal position measurements
+			fuseVelPosHeight(innov[5],innov_var[5],5);
 
-		} else {
-			_time_last_delpos_fuse = _time_last_imu;
+		}else{
+			fuse_mask[VPOS] = false;
+			_innov_check_fail_status.flags.reject_pos_D = true;
 		}
-
-		_innov_check_fail_status.flags.reject_pos_NE = false;
-
-	} else if (!pos_check_pass) {
-		_innov_check_fail_status.flags.reject_pos_NE = true;
 	}
 
-	_fuse_pos = false;
+}
 
-	// record the successful height fusion event
-	if (innov_check_pass_map[5] && _fuse_height) {
-		_time_last_hgt_fuse = _time_last_imu;
-		_innov_check_fail_status.flags.reject_pos_D = false;
+// Helper function that fuses a single velocity or position measurement
+void Ekf::fuseVelPosHeight(const float innov, const float innov_var, const int obs_index)
+{
+	float Kfusion[24] = {}; // Kalman gain vector for any single observation - sequential fusion is used.
+	unsigned state_index = obs_index + 4;	// we start with vx and this is the 4. state
 
-	} else if (!innov_check_pass_map[5]) {
-		_innov_check_fail_status.flags.reject_pos_D = true;
+	// calculate kalman gain K = PHS, where S = 1/innovation variance
+	for (int row = 0; row < _k_num_states; row++) {
+		Kfusion[row] = P[row][state_index] / innov_var;
 	}
 
-	_fuse_height = false;
+	float KHP[_k_num_states][_k_num_states];
 
-	for (unsigned obs_index = 0; obs_index < 6; obs_index++) {
-		// skip fusion if not requested or checks have failed
-		if (!fuse_map[obs_index] || !innov_check_pass_map[obs_index]) {
-			continue;
+	for (unsigned row = 0; row < _k_num_states; row++) {
+		for (unsigned column = 0; column < _k_num_states; column++) {
+			KHP[row][column] = Kfusion[row] * P[state_index][column];
 		}
+	}
 
-		unsigned state_index = obs_index + 4;	// we start with vx and this is the 4. state
+	// if the covariance correction will result in a negative variance, then
+	// the covariance matrix is unhealthy and must be corrected
+	bool healthy = true;
 
-		// calculate kalman gain K = PHS, where S = 1/innovation variance
-		for (int row = 0; row < _k_num_states; row++) {
-			Kfusion[row] = P[row][state_index] / _vel_pos_innov_var[obs_index];
+	for (int i = 0; i < _k_num_states; i++) {
+		if (P[i][i] < KHP[i][i]) {
+			// zero rows and columns
+			zeroRows(P, i, i);
+			zeroCols(P, i, i);
+
+			healthy = false;
+
+			setVelPosFaultStatus(obs_index,true);
```

Comment:
1. It's okay here because it is clear enough that the true/false is the value you want to apply to the obs_index, but good practice is to wrap the function into an other one that sets the boolean for you.
2. I know that there is no astyle check for ecl, but try to follow the rues as close as possible :D 
```suggestion
			setVelPosFaultStatus(obs_index, true);
```